packages/backend-common/src/reading/types.ts

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * A generic interface for fetching plain data from URLs.
+ */
+export type UrlReader = {
+  read(url: string): Promise<Buffer>;
```

Comment:
readFile?

Reply:
Maybe, why? :>

Reply:
As you said we may add more.

The need to read specifically textual content is so common (unique rn) that it could even be readTextFile and return string :)

But no biggie, just nits

Reply:
Yep, really wanna keep this interface lean though, thinking it should be reasonable effort to implement a custom version without ties to `@backstage/backend-common`. Reason I'm hesitant with `readFile` is that we're not really reading files, would rather go `readUrl` in that case. Just `read` is kinda simple though :>

Reply:
I'm assuming this interface could support reading data from any rest endpoint? I'm thinking of the use case where an org has an existing catalog service and want to read from there.

I noticed the githubOrg processor hasn't been moved over. Does that mean the UrlReaders are meant to handle only a subset of entity types?  I guess I'm a little confused on the distinction of what should be a reader and what is a processor in some of these more subtle cases.

Also does this new distinction or reader vs processor cover something like the GitHub GQL API utility that have been previously added into the catalog-backend?

Reply:
This interface isn't actually tied directly to the catalog at all. It is simply something that enables fetching data from URLs, even if it needs to be translated into an authenticated API call.

The way this is then used in the catalog is the `UrlReaderProcessor`, which reads data from locations of type `url`. That in turn replaces the simple processors like `GithubReaderProcessor`, but it doesn't replace the more complex ones like `GithubOrgReaderProcessor`

Reply:
Should also mention that the plan is to expand this interface somewhat, in order for it to be usable by at least the scaffolder and techdocs plugins too

Reply:
Thanks, I think I understand the distinction well enough now. However I could see this being slightly confusion for newcomers to understand the differences and when to use one extension point versus the other.

I think part of the problem is LocationReader's implement different methods that seem to define that distinction, but not always explicitly in their names. It seems like some read and some process, or they do both. I'm not sure it's an issue per say, and I do think this PR definitely helps. Perhaps it's just a matter of having some good docs around this area when it stabilizes.

Reply:
Yep, and short term it's gonna be even more confusing given all the backwards compatibility that's in place 😅 

In general I think we're moving away from many different types of catalog processors to fewer ones. I think we can for example remove the location processing from the catalog, and just replace it all the entity processors that work on location entities, and perhaps also more processors operating directly on the bootstrap location.